biahub/track.py

```diff
@@ -0,0 +1,993 @@
+        A list of input image specifications. Each `ProcessingInputChannel` defines
+        which channels to process and the pipeline (functions + arguments) to apply.
+    visualize : bool, optional
+        If True, opens a Napari viewer to display each processed channel after its pipeline.
+
+    Returns
+    -------
+    dict of str to ArrayLike
+        Updated dictionary where the processed channel data has replaced the originals.
+
+    Notes
+    -----
+    - The `step.input_channels` field can define which channel(s) to use as inputs
+      to a function, even if different from the one being written to.
+    - If `per_timepoint` is True for a step, the function will be applied frame-by-frame
+      using `ultrack.utils.array.array_apply`.
+    - All functions must be registered in the `FUNCTION_MAP` to be resolved.
+
+    Examples
+    --------
+    >>> from biahub.settings import ProcessingInputChannel, ProcessingFunctions
+    >>> import numpy as np
+
+    >>> data_dict = {"raw": np.random.rand(10, 1, 256, 256)}  # shape (T, Z, Y, X)
+
+    >>> input_images = [
+    ...     ProcessingInputChannel(
+    ...         path=None,
+    ...         channels={
+    ...             "raw": [
+    ...                 ProcessingFunctions(
+    ...                     function="np.mean",
+    ...                     kwargs={"axis": 1},
+    ...                     per_timepoint=False,
+    ...                     input_channels=["raw"]
+    ...                 )
+    ...             ]
+    ...         }
+    ...     )
+    ... ]
+
+    >>> output = run_pipeline(data_dict, input_images)
+    >>> output["raw"].shape
+    (10, 256, 256)  # Z-averaged
+    """
+    for image in input_images:
+        for channel_name, pipeline in image.channels.items():
+            for step in pipeline:
+                click.echo(f"Processing {channel_name} with {step.function}")
+                f_name = step.function
+                run_function = resolve_function(f_name)
+                f_kwargs = step.kwargs
+                per_timepoint = step.per_timepoint
+                # if there is input channel, apply the function to the input channel otherwise apply the function to the output channel
+                f_channel_name = step.input_channels
+                if f_channel_name is None:
+                    f_channel_name = [channel_name]
+                f_data = [np.asarray(data_dict[name]) for name in f_channel_name]
+
+                if per_timepoint:
+                    result = array_apply(*f_data, func=run_function, **f_kwargs)
+
+                else:
+                    result = run_function(*f_data, **f_kwargs)
+
+                data_dict[channel_name] = result
+                if visualize:
+                    import napari
+
+                    viewer = napari.Viewer()
+                    viewer.add_image(data_dict[channel_name], name=channel_name)
+
+    return data_dict
+
+
+def read_data(
```

Comment:
This function is used only once in `load_data`, I think we don't need it, i.e. reading the data doesn't need to be that refactored, especially since `shape` and `scale` are not used in `load_data`